File: jump_game.cpp
#include "leetcode.h"

class Solution {
public:
  bool canJump(vector<int> &nums) {
    int minJump = 0;
    for (int i = nums.size() - 2; i >= 0; i--) {
      minJump++;
      if (nums[i] >= minJump)
        minJump = 0;
    }
    if (minJump == 0)
      return true;
    else
      return false;
  }
};

int main() {
  Solution obj;
  vector<int> nums1 = {2, 3, 1, 1, 4}, nums2 = {3, 2, 1, 0, 4};
  testBool(obj.canJump(nums1)); // true
  testBool(obj.canJump(nums2)); // false
}

File: other/jump_game.rs
struct Solution;

impl Solution {
    pub fn can_jump(nums: Vec<i32>) -> bool {
        let mut n = 1;
        for i in (0..nums.len() - 1).rev() {
            n = if nums[i] < n { n + 1 } else { 1 };
        }
        n == 1
    }
}

fn main() {
    let nums1 = vec![2,3,1,1,4];
    let nums2 = vec![3,2,1,0,4];
    println!("{}", Solution::can_jump(nums1));
    println!("{}", Solution::can_jump(nums2));
}
